controllers/storageconsumer/storageconsumer_controller.go

```diff
@@ -52,14 +72,366 @@ func (r *StorageConsumerReconciler) Reconcile(ctx context.Context, request recon
+				r.Log.Info("Failed to remove finalizer from StorageConsumer", "StorageConsumer", klog.KRef(r.storageConsumer.Namespace, r.storageConsumer.Name))
+				return reconcile.Result{}, err
+			}
+		}
+	}
 
 	return reconcile.Result{}, nil
 }
 
+func (r *StorageConsumerReconciler) reconcileCephBlockPool() error {
+
+	storageClusterList := ocsv1.StorageClusterList{}
+	err := r.Client.List(r.ctx, &storageClusterList, client.InNamespace(r.namespace))
+	if err != nil {
+		return fmt.Errorf("Error fetching StorageClusterList. %+v", err)
+	}
+
+	var failureDomain string
+	if len(storageClusterList.Items) != 1 {
+		return fmt.Errorf("Cluster has none or more than 1 StorageCluster")
+	}
+
+	failureDomain = storageClusterList.Items[0].Status.FailureDomain
+
+	capacity := r.storageConsumer.Spec.Capacity.String()
+	desired := &rookCephv1.CephBlockPool{
+		Spec: rookCephv1.NamedBlockPoolSpec{
+			PoolSpec: rookCephv1.PoolSpec{
+				FailureDomain: failureDomain,
+				Replicated: rookCephv1.ReplicatedSpec{
+					Size:                     3,
+					ReplicasPerFailureDomain: 1,
+					TargetSizeRatio:          .49,
+				},
+				Quotas: rookCephv1.QuotaSpec{
+					MaxSize: &capacity,
+				},
+			},
+		},
+	}
+
+	_, err = ctrl.CreateOrUpdate(r.ctx, r.Client, r.cephBlockPool, func() error {
+		if err := r.own(r.cephBlockPool); err != nil {
+			return err
+		}
+
+		r.cephBlockPool.Spec = desired.Spec
+		return nil
+	})
+
+	cephBlockPoolDetails := &ocsv1alpha1.CephResourcesSpec{
+		Kind: r.cephBlockPool.Kind,
+		Name: r.cephBlockPool.Name,
+		CephClients: map[string]string{
+			"provisioner": r.cephClientRBDProvisioner.Name,
+			"node":        r.cephClientRBDNode.Name,
+		},
+	}
+	if r.cephBlockPool.Status != nil {
+		cephBlockPoolDetails.Phase = string(r.cephBlockPool.Status.Phase)
+	}
+	r.storageConsumer.Status.CephResources = append(r.storageConsumer.Status.CephResources, *cephBlockPoolDetails)
+
+	if err != nil {
+		r.Log.Error(err, "Failed to update CephBlockPool.", "CephBlockPool", klog.KRef(r.cephBlockPool.Namespace, r.cephBlockPool.Name))
+		return err
+	}
+
+	return nil
+}
+
+func (r *StorageConsumerReconciler) reconcileCephClientRBDProvisioner() error {
+
+	desired := &rookCephv1.CephClient{
+		Spec: rookCephv1.ClientSpec{
+			Caps: map[string]string{
+				"mon": "profile rbd",
+				"mgr": "allow rw",
+				"osd": fmt.Sprintf("profile rbd pool=%s", r.cephBlockPool.Name),
+			},
+		},
+	}
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.cephClientRBDProvisioner, func() error {
+		if err := r.own(r.cephClientRBDProvisioner); err != nil {
+			return err
+		}
+
+		r.cephClientRBDProvisioner.Spec = desired.Spec
+		return nil
+	})
+
+	cephClientRBDProvisionerDetails := &ocsv1alpha1.CephResourcesSpec{
+		Kind: r.cephClientRBDProvisioner.Kind,
+		Name: r.cephClientRBDProvisioner.Name,
+	}
+	if r.cephClientRBDProvisioner.Status != nil {
+		cephClientRBDProvisionerDetails.Phase = string(r.cephClientRBDProvisioner.Status.Phase)
+	}
```

Comment:
What will be the phase if there is no status on the rook resource?

Reply:
there wont be phase in the cephresources list (ie it will not be defined)